### ISSUE TITLE
Protonmail error: Cannot convert undefined or null to object

### ISSUE THREAD

**Issue by joshirio** [vladimiry/email-securely-app#93]
Recently this error message started to appear on start. What does it mean?

![email-securely-app_2019-01-08_09-09-48](https://user-images.githubusercontent.com/12795018/50817896-a5546500-1325-11e9-9bb3-42a56d96b827.png)

Logs:
```
[2019-01-08 09:08:16.872] [error] [WEBVIEW:protonmail] [lib/database/mail] "messageModel.clearTextBody()" failed on email with the following subject: 16h st 
[2019-01-08 09:08:16.887] [error] [WEB] [reducers/errors] <[account.component][2]> { message: 'Cannot convert undefined or null to object',
  name: 'TypeError' }
```

EDIT: I found the problematic email, it has the subject line `16h st`, but it looks like a normal email just simple text content...

**Comment by vladimiry** [comment]
Printing mail's subject to log has been intentionally enabled as discussed [here](https://github.com/vladimiry/email-securely-app/issues/85#issuecomment-449900983), so now we know on which email it gets failing.

Is there a stacktrace in the log? If not can you run [this build](https://www.dropbox.com/sh/tqisuzqpsvk1gj2/AABZ81ZdUswTvuZH0GnkyoSNa?dl=0&lst=) (it explicitly logs error of this specific case)?

> but it looks like a normal email just simple text content...

Does it mean you can open the email for viewing in browser without errors (clicking on the email in the list)?

**Comment by joshirio** [comment]
Before trying your updated build, I restarted the app and was not able to reproduce the error, it just seems fine now, whereas previously on each start that error message would appear.

Opening the email in question did not cause any error in the logs or UI (tried both views local storage and protonmail webclient). I did not delete or move the problematic email.

Then I downloaded your new build, no errors again... Sorry, let's hope I can reproduce it to get useful info.

**Comment by vladimiry** [comment]
Thanks for the details. I'd recommend using updated build, so if the error occurs again we could get more details (hopefully full stack trace). Let's keep the issue opened.

**Comment by vladimiry** [comment]
> What does it mean?

I didn't answer this question. The error occurs on calling the same Protonmail's `messageModel.clearTextBody()` function as described in [this](https://github.com/vladimiry/email-securely-app/issues/85) issue. The `clearTextBody()` function call is needed to get unencrypted email's body content for putting it into the local database. So there will be no such errors if you disable the local store feature.

The thing is that the app calls `clearTextBody()` for all the messages in the same exact way, without any email content preprocessing and for some emails it just fails. So there is either a need to somehow preprocess email content before calling `clearTextBody()` on it or `clearTextBody()` itself is somehow buggy (inconsistent behaviour).

**Comment by joshirio** [comment]
It did happen again, here's the stack trace:

```
[2019-01-18 09:53:45.390] [error] [WEBVIEW:protonmail] [lib/database/mail] "messageModel.clearTextBody()" failed on email with the following subject: Confirm your Have I Been Pwned registration TypeError: Cannot convert undefined or null to object
    at Function.keys (<anonymous>)
    at Object.getPrivateKeys (webclient1://mail.protonmail.com/app.ce4dd852.js:20270:19)
    at Message.decryptBody (webclient1://mail.protonmail.com/appLazy.module.c338bb90.chunk.js:32897:37)
    at Message._callee4$ (webclient1://mail.protonmail.com/appLazy.module.c338bb90.chunk.js:33083:31)
    at tryCatch (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:179722:40)
    at Generator.invoke [as _invoke] (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:179948:22)
    at Generator.prototype.(anonymous function) [as next] (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:179774:21)
    at asyncGeneratorStep (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:205978:24)
    at _next (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:206000:9)
    at webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:206007:7
    at new Promise (<anonymous>)
    at Message.<anonymous> (webclient1://mail.protonmail.com/vendors~app.60481799.chunk.js:205996:12)
    at Message.clearTextBody (webclient1://mail.protonmail.com/appLazy.module.c338bb90.chunk.js:33142:33)
    at Object.<anonymous> (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:58:5108)
    at Generator.next (<anonymous>)
    at /Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:16:1532
    at new Promise (<anonymous>)
    at Module.l (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:16:1309)
    at Object.assign.body (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:58:5032)
    at Object.<anonymous> (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:58:5245)
    at Generator.next (<anonymous>)
    at /Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:16:1532
    at new Promise (<anonymous>)
    at Module.l (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:16:1309)
    at Object.e.buildMail (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:58:4805)
    at /Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:44:160752
    at Generator.next (<anonymous>)
    at u (/Applications/email-securely-app.app/Contents/Resources/app.asar/app/electron-preload/webview/protonmail.js:16:1365)
[2019-01-18 09:53:46.118] [error] [WEB] [reducers/errors] <[account.component][2]> { message: 'Cannot convert undefined or null to object',
  name: 'TypeError' }
```

I sugned up under "notify me" at https://haveibeenpwned.com/ and then that error occurred. I tried to delete the email and redo the signup to try to reproduce it without success, so this is truly random.

More importantly, the problematic email can be viewed just fine without errors, so not sure what implications this can cause, but if this is something on protonmails end and does not cause significant issues, maybe it would be best to just ignore and hide that message completely, or is it possible to recall the same function on failure maybe the second time it will work fine, I have no idea.

**Comment by vladimiry** [comment]
Thanks for details, can you also specify app version used? So I know which web client version was used. The web client is now built-in and so I can know for sure I get the exact version you were using.

Also, did you happen to notice were there a lot of emails in the syncing queue (like you just started the app after a day or two not using it and so the app started syncing process, blinking icon in green) or just one or few?

> to just ignore

That can be done, but for now, I'd prefer not ignoring things that I don't understand and that is important (decrypting the data, even if just directly calling protonmail's implementation), as that may lead to potential instability. So let's keep the issue unresolved at least for a while.

For example, currently the app ignores some network connection errors and 503/504 errors:
https://github.com/vladimiry/email-securely-app/blob/28d955bfd2afbdffdfd8758e93e434338b67a6c8/src/electron-preload/webview/tutanota/lib/util.ts#L79
https://github.com/vladimiry/email-securely-app/blob/28d955bfd2afbdffdfd8758e93e434338b67a6c8/src/electron-preload/webview/protonmail/lib/uilt.ts#L40
But in this case, a few retry calls will happen with short delay and if it still fails there will be another syncing interaction happening triggered by [backup/timer](https://github.com/vladimiry/email-securely-app/blob/28d955bfd2afbdffdfd8758e93e434338b67a6c8/src/web/src/app/_accounts/accounts.effects.ts#L140) trigger (so there are 3 types of syncing triggers). That sad ignoring, in this case, is happening by design, it's a part of the logic.

> or is it possible to recall the same function on failure maybe the second time it will work fine

If we need to recall the function it will mean that protonmail's implementation is inconsistent as I wrote before:

> clearTextBody() itself is somehow buggy (inconsistent behaviour).

I'm going to take another look into it. If I will be able to detect a pattern then it will make sense to raise a bug here https://github.com/ProtonMail/WebClient

**Comment by joshirio** [comment]
> Also, did you happen to notice were there a lot of emails in the syncing queue (like you just started the app after a day or two not using it and so the app started syncing process, blinking icon in green) or just one or few?

I did not start the app for about 8 hours, and I had maybe 1-2 emails in queue, so nothing special.
After the error, the first thing I did was to restart the app without opening any email, then the error did not reappear again.

**Comment by vladimiry** [comment]
Here is the `clearTextBody()` function call stack (listed JS files belong to https://github.com/ProtonMail/WebClient):
```
src/app/message/factories/messageModel.js
=> clearTextBody() 
    => this.decryptBody()
        =>  keysModel.getPrivateKeys(this.AddressID);
            => src/app/keys/factories/keysModel.js
                => return Object.keys(CACHE[addressID]).reduce((acc, keyID) => {
```

It fails on `Object.keys(CACHE[addressID])` call because cache doesn't contain the data for the key. So it looks like the app needs to wait for keys cache initialization and only then start data syncing process. Keys cache filling happens here `src/app/authentication/services/auth.js => fetchUserInfo()`. Will be exploring the way of introducing a needed delay.

Based on that information the error can potentially occur only at logging into the account stage due to the auth process has not been completely finished yet but the app already started data syncing process which requires keys cache to be initialized. So after getting `Cannot convert undefined or null to object` the first time at logging into the account stage the backup/timer syncing trigger will kick-in in 3 minutes after the original fail and email message will be successfully downloaded as keys cache is initialized this time.

**Comment by joshirio** [comment]
I can confirm this error showed up only on start-up and once. It could be that after a restart the app is still in memory and thus is initialized faster unlike a cold boot scenario.

Good progress.

**Comment by vladimiry** [comment]
> It could be that after a restart the app is still in memory and thus is initialized faster unlike a cold boot scenario.

The app by design [resets](https://github.com/vladimiry/email-securely-app/blob/95c2b033ecd53c4eccd6d2efaee9e57bc0379f4e/src/electron-main/session.ts#L17) all types of cache and browser-based databases on start and on logout actions. So cold boot is a permanent case. It's just a luck that in most of the cases error doesn't occur since needed delay gets caused by some side/parallel activity, so we should thank Electron for being slow when it's necessary :smile:

Here is the [pre 2.2.1 version](https://www.dropbox.com/sh/y413x6c06tt7kbf/AAAtO3tsjsnH40EHVMHpVmy_a?dl=0&lst=) which includes the needed delay. Hopefully it resolves the issue but I'm not clossing it yet. Can you start using this version? Then in case of no new error reports I will publish a new release, I think in a few weeks. This version also includes `504 Gateway Time-out` and `503 ServiceUnavailableError` errors processing which is supposed to supress an annoying errors appearing, those that can be ignored with applying retry logic.

**Comment by joshirio** [comment]
Sure, I'll report if the issue appears again, thanks.

**Comment by vladimiry** [comment]
The [pre 2.2.1](https://www.dropbox.com/sh/y413x6c06tt7kbf/AAAtO3tsjsnH40EHVMHpVmy_a?dl=0&lst=) build has been updated with the fix for [this](https://github.com/vladimiry/email-securely-app/issues/99) issue included.

**Comment by joshirio** [comment]
No issues so far 👍

**Comment by vladimiry** [comment]
Fix has been released with [v2.3.0](https://github.com/vladimiry/email-securely-app/releases/tag/v2.3.0).
